master/internal/cache/cache.go

```diff
@@ -0,0 +1,199 @@
+package cache
+
+import (
+	"context"
+	"io/fs"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/pkg/archive"
+	"github.com/determined-ai/determined/proto/pkg/experimentv1"
+)
+
+type modelDefFolder struct {
+	fileTree   []*experimentv1.FileNode
+	cachedTime time.Time
+	lock       sync.RWMutex
+	path       string
+}
+
+// FileCache is metadata for files cached at file system.
+type FileCache struct {
+	rootDir string
+	maxAge  time.Duration
+	caches  map[int]*modelDefFolder
+	lock    sync.Mutex
+}
+
+// NewFileCache initialize FileCache obj.
+func NewFileCache(rootDir string, maxAge time.Duration) *FileCache {
+	err := os.RemoveAll(rootDir)
+	if err != nil {
+		log.WithError(err).Errorf("failed to clear the content of cache folder at %s", rootDir)
+	}
+	return &FileCache{
+		rootDir: rootDir,
+		maxAge:  maxAge,
+		caches:  make(map[int]*modelDefFolder),
+	}
+}
+
+func (f *FileCache) getOrCreateFolder(expID int) (*modelDefFolder, error) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	value, ok := f.caches[expID]
+	if ok {
+		return value, nil
+	}
+
+	exp := struct {
+		ModelDefinition []byte
+	}{}
+	err := db.Bun().NewSelect().TableExpr(
+		"experiments").Column("model_definition").Where("id = ?", expID).Scan(context.Background(), &exp)
+	if err != nil {
+		return nil, err
+	}
+	var fileTree []*experimentv1.FileNode
+	arc, err := archive.FromTarGz(exp.ModelDefinition)
+	if err != nil {
+		return nil, err
+	}
+	err = os.MkdirAll(f.genPath(expID, ""), 0700)
+	if err != nil {
+		return nil, err
+	}
+	for _, ar := range arc {
+		if ar.IsDir() {
+			err = os.MkdirAll(f.genPath(expID, ar.Path), 0700)
+		} else {
+			err = os.WriteFile(f.genPath(expID, ar.Path), ar.Content, 0600)
```

Comment:
You have to be _really_ careful about unpacking tarballs, because anybody can craft a tarball with arbitrary paths.  So with the implementation you have here, somebody could have a tarball with a path like `../../../etc/shadow`, which they could use to override the root password, then they could become root on the machine.  If the determined service is running as root (right now it does) then that would be trivially possible.

So you either need to check the resulting path first, to ensure that it falls inside `$CACHE/exp_model_def/$EXP_ID`, or you need to avoid writing to user-defined paths at all.

If you wanted to take the second route, your cache directory could be a flat list of uuid-named files, where each uuid corresponds to a single file you've unpacked, and in memory you keep a mapping of paths you've unpacked and the uuid file that it corresponds to.  This would eliminate some of the complexity and danger of your code, but it would make the resulting system a little less nice to deal with.  For instance, if the cache is written as untarred files, like you have now, then any system admin can look though that directory, and our system seems a little less mysterious.

So because of that, I don't care which way you do it.

Reply:
Thank you for catching this, validation for user-defined path has been added